smartsim/experiment.py

```diff
@@ -223,44 +235,87 @@ def execute_dispatch(job: Job) -> LaunchedJobID:
             # it easier to monitor job statuses
             # pylint: disable-next=protected-access
             self._active_launchers.add(launch_config._adapted_launcher)
-            return launch_config.start(exe, env)
+            # Generate the job directory and return the generated job path
+            job_execution_path = self._generate(generator, job, idx)
+            return launch_config.start(exe, job_execution_path, env)
 
-        return execute_dispatch(job), *map(execute_dispatch, jobs)
+        return execute_dispatch(generator, job, 0), *(
+            execute_dispatch(generator, job, idx) for idx, job in enumerate(jobs, 1)
+        )
 
     @_contextualize
-    def generate(
-        self,
-        *args: t.Union[SmartSimEntity, EntitySequence[SmartSimEntity]],
-        tag: t.Optional[str] = None,
-        overwrite: bool = False,
-        verbose: bool = False,
-    ) -> None:
-        """Generate the file structure for an ``Experiment``
-
-        ``Experiment.generate`` creates directories for each entity
-        passed to organize Experiments that launch many entities.
-
-        If files or directories are attached to ``application`` objects
-        using ``application.attach_generator_files()``, those files or
-        directories will be symlinked, copied, or configured and
-        written into the created directory for that instance.
-
-        Instances of ``application``, ``Ensemble`` and ``FeatureStore``
-        can all be passed as arguments to the generate method.
-
-        :param tag: tag used in `to_configure` generator files
-        :param overwrite: overwrite existing folders and contents
-        :param verbose: log parameter settings to std out
+    def _generate(
+        self, generator: Generator, job: Job, job_index: int
+    ) -> os.PathLike[str]:
+        """Generate the directory and file structure for a ``Job``
+
+        If files or directories are attached to an ``application`` object
+        associated with the Job using ``application.attach_generator_files()``,
+        those files or directories will be symlinked, copied, or configured and
+        written into the created job directory
+
+        An instance of ``Generator`` and ``Job`` can be passed as an argument to
+        the protected _generate member, as well as the Jobs index.
+
+        :param generator: The generator is responsible for creating the job run and log directory.
+        :param job: The job instance for which the output is generated.
+        :param job_index: The index of the job instance (used for naming).
+        :returns: The path to the generated output for the job instance.
+        :raises: A SmartSimError if an error occurs during the generation process.
         """
+        # Generate ../job_name/run directory
+        job_path = self._generate_job_path(job, job_index, generator.root)
```

Comment:
This function might be better named ``_generate_run_path``.  Job path makes me think exp_name/job_name/